pvlib/bifacial/infinite_sheds.py

```diff
@@ -0,0 +1,1050 @@
+    tan_psi_top = x1 / x2
+    psi_top = np.arctan2(x1, x2)
+    return psi_top, tan_psi_top
+
+
+def _sky_angle_tangent(gcr, tilt, x):
+    """
+    tangent of angle from a point x along the module slant height to the
+    top of the previous row.
+
+    .. math::
+
+        \\tan{\\psi_{top}} &= \\frac{y \\text{GCR} \\sin{\\beta}}{1 - y
+        \\text{GCR} \\cos{\\beta}} \\newline
+
+        y &= 1 - x
+
+    Parameters
+    ----------
+    gcr : numeric
+        ratio of row slant length to row spacing (pitch)
+    tilt : numeric
+        angle of surface normal from vertical in radians
+    x : numeric
+        fraction of module slant length from module bottom edge
+
+    Returns
+    -------
+    tan_psi_top : numeric
+        tangent of angle from x to top of previous row
+    """
+    #  : \\                    .*\\
+    #  :  \\               .-*    \\
+    #  :   \\          .-*         \\
+    #  :    \\   . .*+  psi_t       \\  previous row
+    #  :     \\.-*__________________ \\
+    #  :       \  ^                    \
+    #  :        \  x                    \
+    #  :         \  v                    \
+    #  :          \<---------P----------->\
+
+    y = 1.0 - x
+    return y * np.sin(tilt) / (1/gcr - y * np.cos(tilt))
+
+
+def _f_sky_diffuse_pv(tilt, tan_psi_top, tan_psi_top_0):
+    """
+    view factors of sky from shaded and unshaded parts of PV module
+
+    Parameters
+    ----------
+    tilt : numeric
+        angle of surface normal from vertical in radians
+    tan_psi_top : numeric
+        tangent of angle from shade line to top of next row
+    tan_psi_top_0 : numeric
+        tangent of angle to top of next row with no shade (shade line at
+        bottom)
+
+    Returns
+    -------
+    f_sky_pv_shade : numeric
+        view factor of sky from shaded part of PV surface
+    f_sky_pv_noshade : numeric
+        view factor of sky from unshaded part of PV surface
+
+    Notes
+    -----
+    Assuming the view factor various roughly linearly from the top to the
+    bottom of the rack, we can take the average to get integrated view factor.
+    We'll average the shaded and unshaded regions separately to improve the
+    approximation slightly.
+
+    .. math ::
+        \\large{F_{sky \\rightarrow shade} = \\frac{ 1 + \\frac{\\cos
+        \\left(\\psi_t + \\beta \\right) + \\cos \\left(\\psi_t
+        \\left(x=0\\right) + \\beta \\right)}{2}  }{ 1 + \\cos \\beta}}
+
+    The view factor from the top of the rack is one because it's view is not
+    obstructed.
+
+    .. math::
+        \\large{F_{sky \\rightarrow no\\ shade} = \\frac{1 + \\frac{1 +
+        \\cos \\left(\\psi_t + \\beta \\right)}{1 + \\cos \\beta} }{2}}
+    """
+    # TODO: don't average, return fsky-pv vs. x point on panel
+    psi_top = np.arctan(tan_psi_top)
+    psi_top_0 = np.arctan(tan_psi_top_0)
+    f_sky_pv_shade = (
+        (1 + (np.cos(psi_top + tilt)
+              + np.cos(psi_top_0 + tilt)) / 2) / (1 + np.cos(tilt)))
```

Comment:
@mikofski I'm confused by the `(1 + cos(tilt)` denominator. Is this function doing something other than computing the average of two view factors described by Eq. 6 of your paper, one from the top of the shaded portion and the second from the bottom of the row?

Reply:
Hi Cliff, The `(1+cos(tilt))` in the denominator refers to Eq. 11 in the paper, because the diffuse POA component already contains the simple single-row view factor `(1+cos(tilt))/2` so I call this `f_sky_pv_shade` an "adjustment factor":

![image](https://user-images.githubusercontent.com/1385621/142360821-eb8f110c-3921-4d42-b54e-01850ecb7f7b.png)

Also, importantly, we found that this "averaging" approach only worked for fixed tilt, and is a **BAD** assumption when the rack tilt approaches horizontal, as it would for trackers at noon, when the _linear_ assumption falls apart. So ignore Eq. 15, 16, 17, & 18. Do I need to publish an _errata_ in the proceedings?

Also more importantly, it's completely unnecessary, because using `np.trapz()` is easy, more accurate, and doesn't require any assumptions. Replace Eq. 15-18 with
```
f_x = np.linspace(0, 1, 100)  # a range of x's
tan_psi_top_fx = _sky_angle_tangent(gcr, tilt, f_x)  # a range of angles corresponding to x
psi_top_x = np.arctan(tan_psi_top_fx)  # a range of angles
y = (1 + np.cos(psi_top_x + tilt)) / (1 + np.cos(tilt)
f_sky_pv_shade = np.trapz(y, x)
```
QED

(ditto for the other part of the module and both vf adjustments for the ground too)

Reply:
Thanks @mikofski I'm going to factor the `1 + cos(beta)` term out so that functions returns the view factor as described. That term is effectively recovering the sky diffuse irradiance from the POA sky diffuse irradiance, and I think that could be a bit more clearly done elsewhere.

I'll also extend the algorithm to return the view factors along a discretization of row height rather than the average. I think it's OK to explain that deviation in code comments.

Reply:
@mikofski this comment in the code is incorrect, I believe:

```
    At the bottom of rack, :math:`x = 0`, the angle is zero, and the view
    factor to the ground is one.
```
It is true that the view factor from the point at the bottom of the to the ground is 1, but an infinitesimal distance up the slant height, it's not 1. So the view factor has a jump discontinuity at `f_x=0` where `f_x` is a fraction of slant height from the bottom. Maybe that's why the two-point approximation didn't turn out well.

Assuming I'm not missing something, I will rework view factor formulas for e.g. the ground view factor to integrate from 0 to x for the shaded portion, without using vf(0) = 1.

Reply:
The two point approximation worked well when the module was tilted at 20-degrees as in the paper. See figures 7, 8, 9, & 10.

![image](https://user-images.githubusercontent.com/1385621/142518429-3c81d83a-9883-4d32-b013-4d199ba511ee.png)

In my opinion, there's no discontinuity. In the figures above, the view factor smoothly goes from 1 to a value infinitesimally less than one at an infinitesimally small distance up the module.

We did not discover the issue until we added trackers, which rotate from east to west and are horizontal in the afternoon. When the modules are horizontal, the view factor is not linear. It's more like a U shape and symmetrical, _EG_: the left and right sides of the module have the same view factor, but the middle is different, so any attempt to approximate it as linear will fail. That's when we realized our error.

I'm fine with splitting the denominator out. That makes it more readable and easier to explain and understand. Thank you for extending the algorithm with the discretized integral. The two point method is wrong and should not be used.

FYI: an alternate method that I'd like to explore one day is crossed strings, which would yield the view factor from one surface to another without integration.

Reply:
I think I understand why we're both right. When I said view factor I mean a fraction of the sky dome (hemisphere); when you say view factor here you mean a fraction of the visible wedge of the dome, i.e., the hemispherical view factor divided by the (1 + cos(beta)) term.

Reply:
I'm sorry. II should clarify, because this statement doesn't make sense. I said:

> It's more like a U shape and symmetrical, EG: the left and right sides of the module have the same view factor, but the middle is different, so any attempt to approximate it as linear will fail. That's when we realized our error.

But actually, I think if the module is horizontal, then ideally its view factor of the ground is one (1) all the way across, so perhaps the two-point approximation would still be valid, but SolarFarmer was cutting some other corners and integrating across the ground as well, which does **not** have a constant view factor of the sky.  So it's really the _product_ of the ground VF _and_ the reflected irradiance, integrated from zero to pi at each point on the module backside that is not linear. The edges of the module are closer to the slice of ground that has the best view, therefore the angles that subtend that brighter more reflective patch of ground are larger than in the middle, where it would be narrower. That's what I meant, but I explained it poorly (and perhaps still am, sorry!).

![image](https://user-images.githubusercontent.com/1385621/142664822-78406504-3059-4a2c-9906-4d93694efd54.png)